src/services/hsm.service.js

```diff
@@ -6,16 +6,208 @@
+     * for the address index, this allows us to use all 2 billion indexes
+     * without resorting to hex conversion since 2 billion in base 10 has
+     * length 10.
+     */
+    (name, opts) => `T1${name}_HAddr_${opts.index}`,
+  ];
+
+  return levels[pathIndex](keyname, options);
+}
+function getAcctKeyname(keyname) {
+  return getBip32Keyname(keyname, 3);
+}
+
+function getChangeKeyname(keyname) {
+  return getBip32Keyname(keyname, 4);
+}
+
+/**
+ * @typedef {Object} SignatureData
+ * @property {Buffer} signature
+ * @property {Buffer} pubkey
+ * @property {number} addressIndex
+ * @property {number} inputIndex
+ */
+
+class HsmSession {
+  /**
+   * @param {hsm.interfaces.Hsm} conn
+   * @param {string} rootKeyname
+   */
+  constructor(conn, rootKeyname) {
+    this.conn = conn;
+    this.rootKey = rootKeyname;
+    /**
+     * @type {Record<number, string>}
+     */
+    this.addressKeys = {};
+  }
+
+  /**
+   * Get the adddress index keyname from the address index
+   * @param {index} index - bip32 Address index
+   * @returns {Promise<string>}
+   */
+  async getAddressKeyName(index) {
+    if (index in this.addressKeys) {
+      return this.addressKeys[index];
+    }
+
+    const addressKeyName = getBip32Keyname(this.rootKey, 5, { index });
+    const changeKeyName = getChangeKeyname(this.rootKey);
+    try {
+      await this.conn.blockchain.createBip32ChildKeyDerivation(
+        hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+        index,
+        false,
+        true,
+        changeKeyName,
+        addressKeyName,
+      );
+    } catch (err) {
+      await delay(500);
+      // (CODE: 5022 | NAME: ERR_OBJ_ALREADY_EXISTS)
+      // This error code means the address key was already derived
+      if (err.errorCode === hsm.constants.ERR_OBJ_ALREADY_EXISTS) {
+        // address key already exists, just return it
+        this.addressKeys[index] = addressKeyName;
+        return addressKeyName;
+      }
+      console.error(err);
+      throw err;
+    }
+
+    this.addressKeys[index] = addressKeyName;
+    return addressKeyName;
+  }
+
+  /**
+   * Sign the given data with the given HSM key
+   * @param {HsmSession} hsmSession
+   * @param {Buffer} dataToSignHash
+   * @param {number} addressIndex
+   * @returns {Promise<Buffer>} DER encoded signature
+   */
+  async signData(dataToSignHash, addressIndex) {
+    // Derive the key to the desired address index
+    const htrKeyName = await this.getAddressKeyName(addressIndex);
+
+    const signature = await this.conn.blockchain.sign(
+      hsm.enums.BLOCKCHAIN_SIG_TYPE.SIG_DER_ECDSA,
+      hsm.enums.BLOCKCHAIN_HASH_MODE.SHA256,
+      dataToSignHash,
+      htrKeyName,
+    );
+
+    // Dinamo SDK returns v + DER, where v is a parity byte
+    // https://manual.dinamonetworks.io/nodejs/enums/hsm.enums.BLOCKCHAIN_SIG_TYPE.html#SIG_DER_ECDSA
+    // We can safely ignore the parity byte
+    if (signature[0] !== 0x30) {
+      return signature.slice(1);
+    }
```

Comment:
So what this is doing is ignoring every byte that is not a sequence byte, can you give more context on this in the comment below?

Also, can 0x30 be a constant?

Reply:
Not every byte, just the first byte if it is not `0x30`.
The `0x30` in DER encoding is a tag byte that marks the subsequent data as a sequence (or array).
Since a DER signature is the `[r,s]` array it should always start with `0x30`, but dinamo adds a parity byte.

We can use the parity byte to check the signature but it is not required.
Our lib (bitcore-lib) will throw an error if the first byte is not `0x30`, so we have to remove it either way.